Tests/Realm.Tests/Database/WriteOverloads.cs

```diff
@@ -0,0 +1,272 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Linq;
+using NUnit.Framework;
+
+namespace Realms.Tests.Database
+{
+    [TestFixture, Preserve(AllMembers = true)]
+    public class WriteOverloads : RealmInstanceTest
```

Comment:
Perhaps it's worth adding a test that ensures that an exception is thrown if the user returns a managed object/query/list indirectly, e.g.:

```csharp
var temp = await _realm.WriteAsync(bgRealm =>
{
    var obj = bgRealm.Add(new IntPrimaryKeyWithValueObject());
    return new
    {
        ManagedObject = obj,
        SomeValue = 123
    };
});

Assert.Throws<*not-sure-what-exception*>(() => temp.ManagedObject.Id);
```